### PR TITLE
Template exceptions fix in database and file importers

### PR DESCRIPTION
**Pull request by Auriga2**
### Prerequisites

- [x] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change
Templates in database and file importers now handle some edge cases and prevent exceptions, specifically when the templates are of an incorrect format. This was specially the case with legacy templates. More notes and discussions are in https://github.com/constellation-app/constellation/issues/1652 after the comment by [GammaVel](https://github.com/GammaVel)  

### Alternate Designs
n/a

### Why Should This Be In Core?
Templates are more usable

### Benefits
Better UX

### Possible Drawbacks
n/a

### Verification Process
As described in detain in https://github.com/constellation-app/constellation/issues/1652 Legacy/incorrect format could be manually created by editing a new template as in the screen captured there. Test both importers with all combinations of options, paying attention to the attributes section. Verify they are updated correctly as per the options saved in the loaded template.

### Applicable Issues
Related to https://github.com/constellation-app/constellation/issues/1652

### CODE REVIEW

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedIO.java [hunk 1]
```diff
@@ -225,6 +227,8 @@ public static void loadParameters(final Window parentWindow, final DelimitedImpo
 
     private static void loadParameterFile(final DelimitedImportController importController, final File delimIoDir,
             final String templName) {
+        List<String> missingUserAttributes = new ArrayList<>();
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedIO.java:230]: final

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedIO.java [hunk 2]
```diff
@@ -303,18 +317,32 @@ private static void loadParameterFile(final DelimitedImportController importCont
...
-                    ((DelimitedImportPane) importController.getStage()).update(importController, definitions);
+                    ((DelimitedImportPane) importController.getImportPane()).update(importController, definitions);
                 } finally {
                     importController.setClearManuallyAdded(true);
                 }
             } else {
                 final String message = String.format("Can't find schema factory '%s'", destination);
                 NotifyDisplayer.displayAlert(LOAD_TEMPLATE, "Destination Schema Error", message, Alert.AlertType.ERROR);
             }
+        } catch (final NullPointerException ex) {
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedIO.java:341]: I think my preference would be to do null checks over catching a null pointer as it's better practice that way. Sonar will also be flagging this. Potentially what you could is do a null check and then throw a different kind of exception (probably runtime) in the else block

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/ImportJDBCIO.java [hunk 3]
```diff
@@ -273,19 +292,32 @@ private static void loadParameterFile(final JDBCImportController importControlle
...
                     importController.setClearManuallyAdded(true);
                 }
             } else {
-                final String msg = String.format("Can't find schema factory '%s'", destination);
-                final NotifyDescriptor nd = new NotifyDescriptor.Message(msg, NotifyDescriptor.ERROR_MESSAGE);
-                DialogDisplayer.getDefault().notify(nd);
+                final String message = String.format("Can't find schema factory '%s'", destination);
+                NotifyDisplayer.displayAlert(LOAD_TEMPLATE, "Destination Schema Error", message, Alert.AlertType.ERROR);
             }
+        } catch (final NullPointerException ex) {
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/ImportJDBCIO.java:316]: similar here

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/ImportJDBCIO.java [hunk 4]
```diff
@@ -211,6 +218,8 @@ public static void loadParameters(final Window parentWindow, final JDBCImportCon
     }
 
     private static void loadParameterFile(final JDBCImportController importController, final File delimIoDir, final String templName) {
+        List<String> missingUserAttributes = new ArrayList<>();
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/ImportJDBCIO.java:221]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedIO.java (hunk 2 above)
Comment (Auriga2) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedIO.java:341]: I tried to avoid checking null in ~17 lines which are unlikely/less likely to occur. NullPointerException is thrown at the JsonNode.get() hence cannot throw another exception and handle it in this code.

**Comment by Auriga2** [comment]
@antares1470 I think I fixed it as you mentioned. Also refactored the class by moving the common code to a utility class. (Ideally the `saveParameters` could also be refactored to reduce the Cognitive Complexity)

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                    + " manually or use a new template.", fieldName);
+            NotifyDisplayer.displayAlert(LOAD_TEMPLATE, "Template Error", message, Alert.AlertType.ERROR);
+            throw new NoSuchElementException(message);
+        }
+    }
+
+    public static List<ImportDefinition> getImportDefinitions(final ImportController importController, final JsonNode root, final String templName) {
+        final List<ImportDefinition> definitions = new ArrayList<>();
+        int tabCount = 0;
+        final JsonNode source = TemplateUtilities.getRequiredFieldFromTemplate(root, SOURCE);
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java:76]: you should be able to call the function without the explicit call of `TemplateUtilities` since you are still operating in that class. Similarly for 2 of the next 3 lines after this one

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        return definitions;
+    }
+
+    private static RowFilter populateFilter(final JsonNode definitionNode) {
+        final RowFilter filter = new RowFilter();
+        if (definitionNode.has(FILTER)) {
+            final JsonNode filterNode = TemplateUtilities.getRequiredFieldFromTemplate(definitionNode, FILTER);
+            final String script = TemplateUtilities.getRequiredFieldFromTemplate(filterNode, SCRIPT).textValue();
+            final JsonNode columnsArray = filterNode.withArray(COLUMNS);
+            final ArrayList<String> columns = new ArrayList<>();
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java:118]: Can you make the left hand side just a `List<String>`

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java (hunk 5 above)
Comment (Auriga2) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java]: Oh initially added in ImportDelimitedIO.java. A copy paste issue! Fixed now.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java (hunk 6 above)
Comment (Auriga2) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java]: Just copy pasted, fixed now.

**Review by antares1470 — APPROVED** [review]

**Comment by mimosa2** [comment]
@Auriga2, A minor update - include author & class details for TemplateUtilities.java.

**Review by mimosa2 — APPROVED** [review]

**Review by mimosa2 — APPROVED** [review]
